lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+    first_t: u32,
+    tstamps_mem: &mut [TimeStamp; 2],
+    phi: u32,
+    fscale: u32,
+    tadc: u32,
+) -> [f32; SAMPLE_BUFFER_SIZE] {
+    let overflow_count: u32 = tadc * SAMPLE_BUFFER_SIZE as u32;
+    let tref_count: u32 = tstamps_diff(tstamps_mem, overflow_count);
+    let mut thetas: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
+    let mut theta_count: u32;
+
+    // 68ns
+    if tstamps_mem[0].sequences_old == 0 {
+        theta_count = tref_count - first_t;
+    } else {
+        theta_count = tstamps_diff(
+            &[
+                TimeStamp {
+                    count: 0,
+                    sequences_old: 0,
+                },
+                tstamps_mem[0],
+            ],
+            overflow_count,
+        );
+    }
+
+    thetas[0] = real_phase(theta_count, fscale, tref_count, phi);
+    for i in 1..SAMPLE_BUFFER_SIZE {
+        theta_count += tadc;
+        thetas[i] = real_phase(theta_count, fscale, tref_count, phi);
+    }
+
+    thetas
+}
+
+/// Number of fast clock counts between two consecutive
+/// TimeStamps. This requires that `tstamps[0]` is more recent than
+/// `tstamps[1]` but otherwise imposes no restrictions on them. For
+/// instance, they can be from different processing periods and this
+/// will still count the number of counts between them, accounting for
+/// overflow wrapping. Both TimeStamps must be valid for this to be
+/// used.
+///
+/// # Arguments
+///
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `overflow_count` - Max timestamp value.
+fn tstamps_diff(tstamps_mem: &[TimeStamp; 2], overflow_count: u32) -> u32 {
+    debug_assert!(tstamps_valid_count(tstamps_mem) == 2);
+
+    if tstamps_mem[0].sequences_old == tstamps_mem[1].sequences_old {
+        return tstamps_mem[0].count - tstamps_mem[1].count;
+    }
+
+    let rem0: u32 = tstamps_mem[0].count;
+    let rem1: u32 = overflow_count - tstamps_mem[1].count;
+    let empty_sequences =
+        tstamps_mem[1].sequences_old - tstamps_mem[0].sequences_old - 1;
+
+    rem0 + rem1 + overflow_count * empty_sequences as u32
+}
+
+/// Increment `sequences_old` in each TimeStamp of `tstamps_mem`.
+fn increment_tstamp_sequence(tstamps_mem: &mut [TimeStamp; 2]) {
+    for tstamp in tstamps_mem {
+        tstamp.new_sequence();
+    }
+}
+
+/// Compute the phase (in radians) for a phase given in counts
+/// relative to some period in counts.
+///
+/// # Arguments
+///
+/// * `theta_count` - Phase in counts. This can be greater than the
+/// period in counts.
+/// * `fscale` - Frequency scaling factor for harmonic demodulation.
+/// * `period_count` - Number of counts in 1 period.
+/// * `phase_count` - Phase offset in counts.
+///
+/// # Latency: 138ns
+fn real_phase(
+    theta_count: u32,
+    fscale: u32,
+    period_count: u32,
+    phase_count: u32,
+) -> f32 {
+    let total_angle =
+        modulo::<u32>(theta_count * fscale + phase_count, period_count);
+    2. * PI * (total_angle as f32 / period_count as f32)
```

Comment:
Would you mind further explaining the use of integer phase here? It seems to me that this would only be possible if the external reference period were fixed and *known ahead of time*. This would allow creating a LUT with the same number of indices as the reference period (worth noting: this has the downside of requiring large LUTs for slow reference frequencies). However, I was under the impression that the external reference period is unknown and must be computed at runtime based on the timestamps. Can I treat the reference period as a compile-time input? Or, is there another way to accomplish this result using integer phase that I'm not considering?

Reply:
Several things:

* It doesn't make sense to ever do f32 modulo since `sin()` and `cos()` (when implemented correctly) will do that.
* Integer modulo is always exact. Contrast that with float modulo (i.e. convert to float first and then do modulo something) where you kill accuracy quickly.
* Take a look at the two cases and work through the math. One of them has know and fixed reference period.
* You don't need a big LUT. Interpolation is cheap (compared to the polynomial approach in the existing impls). Math can be used to estimate LUT size from accuracy requirement.
* Also since we know that `sin/cos` will be a performance factor, just take a look around how that's implemented usually when performance is key and accuracy is given. E.g. cmsis dsp has a simple 512-LUT with linear interpolation. That also matches what would typically be implemented in FPGAs.

Reply:
Also I am a bit uncertain about the cmsis quality and I don't understand some design decisions there either. Why don't they remap into the first quadrant (is it really better to save a few sign flips an pay for 4 times the storage)? And they don't seem to handle values outside [0, 2pi) at all.
If I were to implement it with a LUT, I'd try copying the usual [FPGA implementation](https://github.com/m-labs/misoc/blob/master/misoc/cores/cossin.py). But I'm a bit inexperienced to predict whether this is faster than the polynomial approach.
From what I can see, because the phase handling should probably be integer, we'd want `cossin(i32) -> (f32, f32)`.
And in general I'd really like to know whether using integer math for all the DSP (not just phases/PLL) might be better alltogether. But let's just keep that in mind, not act on it now.